### PR TITLE
Add script to convert leaflet-providers.js provider definitions to JSON / generate python file

### PR DESCRIPTION
**Pull request by jorisvandenbossche**
Related to https://github.com/darribas/contextily/issues/10

Starting on a more extensive set of tile providers: started with parsing the leaflet-providers. Still need to do some further tweaking of the resulting JSON (process attributions for javascript placeholders, rename some keys to be more pythonic).

Then, the more important question is how to expose this in the python contextily interface? I could do something similar as https://github.com/jupyter-widgets/ipyleaflet/blob/master/ipyleaflet/basemaps.py, automatically creating such objects from the parsed JSON. 

Another question is how the user can override certain keys, if needed (eg some need an access token or api key, see some examples in javascript here https://github.com/leaflet-extras/leaflet-providers#here-formerly-nokia)

### CODE REVIEW

**Comment by jorisvandenbossche** [comment]
Updated the script to also generate a python file, which I committed to the repo. Some feedback would be good now. 

Still need to integrate those providers in the rest of the package (constructing the correct url from those dictionaries). 
And we also might want to only take a subset for now (eg there are a bunch of them that are overlay tiles, which doesn't work yet with contextily I think).

**Comment by jorisvandenbossche** [comment]
Any feedback on this?

**Comment by ljwolf** [comment]
Gotcha, sorry, must've missed this w/o a tag to review. the generation code looks fine, but I think we still need to figure out how to remove/populate all the extra options. For instance, take:
```python
from contextily._providers import providers
providers.CartoDB['Positron']
```
Those subdomains & resolutions need to be parsed & inserted... not sure how to do this in general?

**Comment by jorisvandenbossche** [comment]
> Those subdomains & resolutions need to be parsed & inserted... not sure how to do this in general.

What do you mean exactly?

I get for this:

```
In [5]: providers.CartoDB['Positron'] 
Out[5]: 
{'url': 'https://{s}.basemaps.cartocdn.com/{variant}/{z}/{x}/{y}{r}.png',
 'attribution': '&copy; <a href="https://www.openstreetmap.org/copyright">OpenStreetMap</a> contributors &copy; <a href="https://carto.com/attributions">CARTO</a>',
 'subdomains': 'abcd',
 'max_zoom': 19,
 'variant': 'light_all',
 'name': 'CartoDB.Positron'}
```

so for constructing the url, we can use the other keys to fill in the variables (in this case only 'variant', but sometimes it will be more, like api keys), and I think we only need to special case 's' (use eg always the first character of the subdomains) and 'r' (I would need to look up what this should be, I think it is for retina resolution, so for now we can default to empty string and enable optional retina later)

**Comment by ljwolf** [comment]
Right, sorry, wasn't specific enough. Yes, that's exactly what I meant. I'm not yet sure if the first character of those `subdomains` strings will always be valid. We might need to test that, but it's not a blocker!

**Comment by jorisvandenbossche** [comment]
If it is not valid, that seems like a bug in the leaflet-providers spec, as I suppose they use that as well (we could see what their logic is)

**Comment by jorisvandenbossche** [comment]
This is ready for testing (basic functionality to construct the url from the dict is implemented). 

Using the max_zoom and attribution text can be left for later I think.

**Comment by jorisvandenbossche** [comment]
@ljwolf @darribas could someone do a basic review of this?

Some items:

- we should probably include *some* manual vetting of the list of providers (eg for now leave out the overlay tiles)
- Are we happy with the user interface? 
  Currently it is:

  ```
    import contextily
    contextily.add_basemap(ax, url=contextily.providers.Stamen.Toner)
    contextily.add_basemap(ax, url=contextily.providers.HERE.NormalDay(app_id=...))
  ```

  but:

  * do we want to provide string aliases? So you can do `contextily.add_basemap(ax, url="Stamen.Toner")` to have something shorter?
  * do we want to keep the `url` keyword? Since we are changing this functionality, it might be an opportunity to change that as well. My main "gripe" is that is both used to pass a url (although the provider dict is now more than a url) or a local file path. We could rename it, or provide two keywords for the two functionalities, ..

* We can start using the attribution text, check the max_zoom, check the bounds, ec based on the metadata of a provider. But I suppose this part can be done in a separate PR.

**Comment by darribas** [comment]
OK I've had a spin and I think this looks good. A couple of (minor) things before merging and we're good to go:

- [ ] Given this is quite a revamp of how we manage where to get data from the web, I suggest a notebook is included explaining the basic functionality and demo'ing it
- [ ] Event though it's not wrong, the docstrings (e.g. `bounds2img`) could be a bit more descriptive as to what `url` can take and what the default is. This would help users get a sense of how to jump on board with the new system

File: contextily/tile.py [hunk 1]
```diff
@@ -22,7 +22,7 @@
 
 
 def bounds2raster(w, s, e, n, path, zoom='auto',
-                  url=sources.ST_TERRAIN, ll=False,
+                  url=None, ll=False,
```
Comment (darribas) [inline contextily/tile.py]: This would need to get updated on the docstring.

File: contextily/tile.py [hunk 2]
```diff
@@ -99,7 +99,7 @@ def bounds2raster(w, s, e, n, path, zoom='auto',
 
 
 def bounds2img(w, s, e, n, zoom='auto',
-               url=sources.ST_TERRAIN, ll=False,
+               url=None, ll=False,
```
Comment (darribas) [inline contextily/tile.py]: Same here for its docstring

**Comment by ljwolf** [comment]
As we were talking elsewhere, one way to get around the transition is to try to come up with a "clean" representations of the new providers. This would mean that we could call 
`str(provider)` and get a valid URL that can be populated with the right `{z}/{x}/{y}.png` information. One way to do this would be to define the `__repr__` or `__str__` method for the `TileProvider`. I've tried:
```python
    def __str__(self):
        content = self.copy()
        content.update({'s':content.get('subdomains', [''])[0]})
        content.update({'z':'{z}', 'x':'{x}', 'y':'{y}'})
        content.setdefault('r', '')
        # workaround for NASAGIBS
        content.setdefault('maxZoom', content.get('max_zoom', None))
        return self.url.format(**content)
```
`NASAGIBS` uses `maxZoom` instead of `max_zoom` in the URL string, but beyond that, this strategy works to provide a URL populated with the content in the `MapProvider` default.

But, for some reason, the `self.copy()` from `__call__` returns an object of type `dict`, not of type `TileProvider`. So... this works only if the URL isn't updated using the call method. I'd like to figure that out, though, cause having this `__str__` method would definitely make it easier to parse/work with these.

**Comment by ljwolf** [comment]
Overriding the copy method for `TileProviders` makes it possible to populate fields & then get the url in the str representation:
```python
def copy(self):
    return TileProvider(**{k:v for k,v in self.items()})
```
so then the update works:
```python
import contextily
str(contextily.providers.Thunderforest.Transport(apikey='my_key'))
# outputs: 'https://.tile.thunderforest.com/transport/{z}/{x}/{y}.png?apikey=my_key'
```

**Comment by jorisvandenbossche** [comment]
@ljwolf I think providing this `str` representation might be useful in general (to get the full url with only x/y/z to fill in), but I don't fully understand how that helps with the transition (because that str is the old format, which is still accepted).

I wouldn't use this for `repr`, as it would otherwise get difficult to see that this is a dict-like `TileProvider` object.

> NASAGIBS uses maxZoom instead of max_zoom in the URL string

Ah, that seems a bug. I think in the dict keys I "pythonized" maxZoom to max_zoom, but I was assuming this was never used in the urls. So I have to update it in the urls as well.

> But, for some reason, the self.copy() from __call__ returns an object of type dict, not of type TileProvider

Ah, that's another good catch. The `__call__` should still result in such a TileProvider subclass. Will fix that and add tests for it.

**Comment by darribas** [comment]
One more thought I had somewhere else, to bring to the discussion here: would it make sense to add a new argument `provider` or `prov` instead of bending the `url` one to take not only strings?

Pros:

- Clean, parallel approach to the current API, which will always work in the same ways it's done (that's still the case with the current proposal, but `url` would *only* take strings on this model).
- More intuitive say when you're reading local files, to think of `url` as the one to pass paths rather than to potentially pass other objects
- (Probably the main reason for me) I find it really unintuitive that `url` can take a string or a dict that, among other things, has a `url` with a string but also a bunch of other objects (zoom, attribution, etc.) all of them also existing as attributes.

Cons:

- New argument. But this one is not that bad I think, it sends a clearer message that you can specify all the required arguments (`url`, `zoom`, `attribution`) on your own, _or_ pass a `provider` object (dict) with all of them there, which might come straight from `providers` or you build it yourself if it's a more bespoke service.

What do y'all think about this?

**Comment by jorisvandenbossche** [comment]
If we create new keywords, instead of splitting it based on the type like string (url) vs dict (provider), I would rather split it for functionality: online tile provider vs single local file. 
In the end, a single url string could be seen as "just" a short-hand for passing `{'url': url_string}` that can be easily supported, also in the new provider keyword. While the local file path is really different: it signifies a local file (not url of tile provider) and a single file (not a bunch of tiles).

**Comment by darribas** [comment]
OK I see your point. Maybe a useful exercise here, if we're going to break backwards compatibility (which is fine for 1.0) would be to think what the ideal API would be if we didn't have the legacy we have?

I see several changes/options:

- Local files have their own attribute (maybe `path`?)
- Online tiles use `url`, `zoom`
- `attribution` can be used for everyone
- Both online tiles _and_ local files can also be specified with a `provider` object, which is a `dict` that contains all the arguments needed. This would allow having a `url` or a `path` and using the `attribution`. Arguably, it'd be more useful for online providers but could serve a purpose for local files (maybe we could stuff some of those into a `.tif` when saved?). At the minimum, I could see myself creating that `dict` for a file I was going to use very often and saving it as a `.json` for reuse...

More options? Variants?

**Comment by jorisvandenbossche** [comment]
I added some extra tests, so coverage is not decreasing anymore now.

**Pull request merged**